packages/contracts-bedrock/contracts/EAS/EAS.sol

```diff
@@ -0,0 +1,853 @@
+        if (address(resolver) == address(0)) {
+            // Ensure that we don't accept payments if there is no resolver.
+            for (uint256 i = 0; i < length; i = uncheckedInc(i)) {
+                if (values[i] != 0) {
+                    revert NotPayable();
+                }
+            }
+
+            return 0;
+        }
+
+        uint256 totalUsedValue = 0;
+
+        for (uint256 i = 0; i < length; i = uncheckedInc(i)) {
+            uint256 value = values[i];
+
+            // Ensure that we don't accept payments which can't be forwarded to the resolver.
+            if (value != 0 && !resolver.isPayable()) {
+                revert NotPayable();
+            }
+
+            // Ensure that the attester/revoker doesn't try to spend more than available.
+            if (value > availableValue) {
+                revert InsufficientValue();
+            }
+
+            // Ensure to deduct the sent value explicitly and add it to the total used value by the batch.
+            unchecked {
+                availableValue -= value;
+                totalUsedValue += value;
+            }
+        }
+
+        if (isRevocation) {
+            if (!resolver.multiRevoke{ value: totalUsedValue }(attestations, values)) {
+                revert InvalidRevocations();
+            }
+        } else if (!resolver.multiAttest{ value: totalUsedValue }(attestations, values)) {
+            revert InvalidAttestations();
+        }
+
+        if (last) {
+            _refund(availableValue);
+        }
+
+        return totalUsedValue;
+    }
+
+    /**
+     * @dev Calculates a UID for a given attestation.
+     *
+     * @param attestation The input attestation.
+     * @param bump A bump value to use in case of a UID conflict.
+     *
+     * @return Attestation UID.
+     */
+    function _getUID(Attestation memory attestation, uint32 bump) private pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    attestation.schema,
+                    attestation.recipient,
+                    attestation.attester,
+                    attestation.time,
+                    attestation.expirationTime,
+                    attestation.revocable,
+                    attestation.refUID,
+                    attestation.data,
+                    bump
+                )
+            );
+    }
+
+    /**
+     * @dev Refunds remaining ETH amount to the attester.
+     *
+     * @param remainingValue The remaining ETH amount that was not sent to the resolver.
+     */
+    function _refund(uint256 remainingValue) private {
+        if (remainingValue > 0) {
+            // Using a regular transfer here might revert, for some non-EOA attesters, due to exceeding of the 2300
+            // gas limit which is why we're using call instead (via sendValue), which the 2300 gas limit does not
+            // apply for.
+            payable(msg.sender).sendValue(remainingValue);
```

Comment:
Using our safecall library is safer and cheaper than `sendValue` because it is vulnerable to returndata bombs, altho in this case there is no incentive to do so

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/121be5dd09caa2f7ce731f0806b0e14761023bd6/contracts/utils/Address.sol#L41